## CREDIT CARD FRAUD DETECTION USING MACHINE LEARNING

Importing the Dependencies

(The libraries and functions we need.)

# NUMPY

Numpy is used for making arrays.

# PANDAS

We have our dataset in csv file.

csv file contains comma separated values

In order to process and analyse the data from a csv file, we need pandas library.

Pandas is used to make dataframes.

dataframes are structured tables.

We feed our data to pandas dataframe and then do the further analysis and processing.

## SKLEARN
scikit-learn

train_test_split function allows us to split our data into training data and testing data

accuracy_score function helps us to check the performance of our model

In [178]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

Loading the dataset to a Pandas Dataframe

pandas.read_csv(): to read csv file loaded to our pandas dataframe

In [179]:
credit_card_data = pd.read_csv('creditcard.csv')

Printing the first 5 rows of the dataset

The head() function shows the 1st 5 rows of the dataframe, and the tail() function shows the last 5 rows of the dataframe.

The 1st row shows the time of transaction in seconds

other details are some confidential/ sensitive credit card details/ features that cannot be exposed

The dataset provider has converted all the features through principle component analysis method, which will convert all the features to some numerical values.

We will be using these numerical values for our analysis and prediction.

The last second column shows amount expressed in US Dollars.

last column class shows whether the particular transaction is legit or fraudulent.

There are 2 classes or labels:

0 represents the transaction is legit transaction

1 represents that the trasaction is a fradulent transaction

In [180]:
credit_card_data.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

The dataset contains about 284806 data points.

In [181]:
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

Dataset Informations.

It tells the number of enteries, no. of columns, datatypes of each of the columns, no.of values present.

The dataset is unbalanced.

In [182]:
credit_card_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

Checking the number of missing values in each column

We don't have any missing values.

In [183]:
credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

Checking the distribution of legit transaction and fradulent transaction.

0 - legit transaction

1 - fradulent transaction

We have 284315 legit transactions and 492 fradulent transaction

In [184]:
credit_card_data['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

If we train our model with this dataset the model accuracy will not be good as a good percentage of dataset belongs to one particular class. So, our model won't be able to easily predict fradulent credit card transactions.
Therefore this dataset is an unbalanced dataset.

This dataset is highly unbalanced.

0 - legit/ normal transaction

1 - fradulent transaction

Now, we will be separating normal transaction and fradulent transaction.

Separating the dataset for analysis.

In [185]:
legit = credit_card_data[credit_card_data.Class == 0]
fraud = credit_card_data[credit_card_data.Class == 1]

We will now print the shape of this pandas series.

In [186]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


Statistical Measures of the data

count value tells the number of data points

mean gives the mean value of amount

25% means that 25% of data is less than 5.6500

In [187]:
legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [188]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

Compare the values for both transactions

gives the mean of all the column after grouping it by Class

In [189]:
credit_card_data.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

Dealing with unbalanced data

Under-Sampling: Build a sample dataset from the original dataset containing similar distribution of legit transactions and fradulent transactions.

Since we know that the number of fradulent transactions is 492.

So, out of 284315 legit transactions we will take random 492 transactions and then join it with the fradulent transactions.

So, the final dataset will have equal number of legit and fradulent transactions.

This dataset will have uniform distribution of legit and fradulent datasets. This will help in doing better predictions using our machine learning model.

In [190]:
legit_sample = legit.sample(n = 492)

Concatenating two Dataframes

If axis is 0 dataset is added one by one row wise.

Axis 0 - row

Axis 1 - column

In [191]:
new_dataset = pd.concat([legit_sample, fraud], axis = 0)

In [192]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
117018   74512.0 -1.195103  1.318795  0.407691  0.964365 -0.497160 -0.382638   
206527  136243.0  2.037688 -0.155753 -1.759825  0.263751  0.363801 -0.703538   
202750  134475.0 -1.116022 -0.243028  0.109692 -0.683660  0.750476 -0.560708   
60046    49131.0  0.092937  1.538071  0.757766  2.799238  0.697949 -0.564306   
17007    28360.0  1.170407  0.179052  0.556120  0.722913 -0.495708 -0.836266   

              V7        V8        V9  ...       V21       V22       V23  \
117018  0.062566  0.809988 -0.788482  ...  0.271128  0.507071 -0.045282   
206527  0.287437 -0.166234  0.270445  ...  0.005232  0.084079  0.111188   
202750 -0.012576  0.601766 -0.019510  ...  0.048794 -0.303187 -0.020429   
60046   1.194055 -0.514176 -1.441387  ...  0.114385  0.714501  0.021460   
17007   0.044742 -0.152274 -0.031150  ...  0.115578  0.416438  0.005372   

             V24       V25       V26       V27       V28  Amount  Class  
117018  0.071192 -0.336066 -0.310862 -0.155988 -0.023998   52.35      0  
206527  0.794979  0.122887  0.507701 -0.114104 -0.080491   12.95      0  
202750  0.698218  0.169246 -0.138411 -0.157694 -0.234615   37.73      0  
60046   0.437878 -1.013480  0.053279  0.200814  0.085683    1.13      0  
17007   0.664414  0.390533  0.423425 -0.016731  0.015406   12.04      0  

[5 rows x 31 columns]

In [193]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [194]:
new_dataset['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [195]:
new_dataset.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      93974.235772 -0.093224  0.022487 -0.043503 -0.068475  0.016463   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.061694 -0.007305 -0.039912 -0.038928  ... -0.025770 -0.023046   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.023014 -0.030238  0.053831 -0.016918  0.002759 -0.019713 -0.012822   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       93.146402  
1      122.211321  

[2 rows x 30 columns]

Splitting the data into features and targets.

X - features

Y - labels

In [196]:
X = new_dataset.drop(columns = 'Class', axis = 1)
Y = new_dataset['Class']

In [197]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
117018   74512.0 -1.195103  1.318795  0.407691  0.964365 -0.497160 -0.382638   
206527  136243.0  2.037688 -0.155753 -1.759825  0.263751  0.363801 -0.703538   
202750  134475.0 -1.116022 -0.243028  0.109692 -0.683660  0.750476 -0.560708   
60046    49131.0  0.092937  1.538071  0.757766  2.799238  0.697949 -0.564306   
17007    28360.0  1.170407  0.179052  0.556120  0.722913 -0.495708 -0.836266   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [198]:
print(Y)

117018    0
206527    0
202750    0
60046     0
17007     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


Splitting the data into training data and testing data

We are craeting 4 arrays X_train, X_test, Y_train, Y_test

test_size = 0.2 indicates that 20% of data will be used for testing and 80% of data will will be used for training.

We want to startify the data based on Y.

In Y the distributionof 0 and 1 can be very different in training data and testing data, using startify there will be evenly distributed 2 classes in both x_train and X_test.

random_state is just to randomize dataset in different ways.

In [199]:
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, stratify = Y, random_state=2)

In [200]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


We had 984 values in our original dataset, about 80% that is 787 values go to X_train and 20% that is about 197 values go to Y_train.

Model Training

We use Logistic Regression for Binary Classification problems.

We are loading a instance of Logistic Regression model in a variable model.

In [201]:
model = LogisticRegression()

Training the Logistic Regression Model with training data.

model.fit is used to fit our data according to the Logistic Regression Model.

In [202]:
model.fit(X_train, Y_train)

LogisticRegression()

Model Evaluation

Checking the performance of your model based on the accuracy score

Accuracy Score

Accuracy Score on Training data

We will predict the labels of the training data and store the predicted labels in X_train_prediction variable.

Then, we will compare the values predicted by our model (stored in X_train_prediction) with the original label values (stored in Y_train)

It will compare the values and give us the accuracy score and the accuracy score will be stored in training_data_accuracy variable.

In [203]:
X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, Y_train)

In [204]:
print('Training data accuracy: ', training_data_accuracy)

Training data accuracy:  0.9415501905972046


Accuracy Score on Testing data

In [205]:
X_test_prediction = model.predict(X_test)
testing_data_accuracy = accuracy_score(X_test_prediction, Y_test)

In [206]:
print('Testing data accuracy: ', testing_data_accuracy)

Testing data accuracy:  0.9289340101522843


If the model is overfitted or under-fitted, the training and testing accuracy have a huge difference.

If the testing data accuracy is too large as compared to testing data accuracy, the model is over-fitted.

If the testing data accuracy is much less as compared to traing data accuracy, the model is under-fitted.